tracker/derived.go

```diff
@@ -0,0 +1,235 @@
+package tracker
+
+import (
+	"github.com/vsariola/sointu"
+	"iter"
+	"slices"
+)
+
+/*
+	from modelData we can derive useful information that can be cached for performance
+	or easy access, because of nested iterations over the score or patch data.
+	i.e. this needs to update when the model changes, and only then.
+*/
+
+type (
+	derivedForUnit struct {
+		unit       *sointu.Unit
+		instrument *sointu.Instrument
+		sends      []*sointu.Unit
+	}
+
+	derivedForTrack struct {
+		instrumentRange          []int
+		tracksWithSameInstrument []int
+		title                    string
+	}
+
+	derivedModelData struct {
+		// map unit by ID
+		forUnit map[int]derivedForUnit
+		// map track by index
+		forTrack map[int]derivedForTrack
+	}
+)
+
+// public access functions
+
+func (m *Model) forUnitById(id int) *derivedForUnit {
+	forUnit, ok := m.derived.forUnit[id]
+	if !ok {
+		return nil
+	}
+	return &forUnit
```

Comment:
This causes heap escaping because forUnit is a local variable and taking a pointer to a local variable requires moving the variable to heap before returning. One nuance of maps is that you cannot take address of a map value i.e. &m["key"] is an error. Thus, the return value of this probably should be derivedUnit, not *derivedUnit, and you might consider returning ok too in that case to indicate if the struct was populated.

Reply:
To be clear: returning a small value type is usually faster than doing a heap allocation, and returning also ok or error to indicate if the struct is populated is idiomatic go

Reply:
should be implemented thoroughly now (in derived.go)